engine/execution/computation/computer/uploader/uploader.go

```diff
@@ -0,0 +1,128 @@
+package uploader
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"path"
+
+	"cloud.google.com/go/storage"
+	"github.com/onflow/flow-go/engine"
+	"github.com/onflow/flow-go/engine/execution"
+	"github.com/onflow/flow-go/module"
+	"github.com/rs/zerolog"
+)
+
+type Uploader interface {
+	Upload(computationResult *execution.ComputationResult) error
+}
+
+func NewAsyncUploader(uploader Uploader, log zerolog.Logger, metrics module.ExecutionMetrics) *AsyncUploader {
+	return &AsyncUploader{
+		unit:     engine.NewUnit(),
+		uploader: uploader,
+		log:      log.With().Str("component", "block_data_uploader").Logger(),
+		metrics:  metrics,
+	}
+}
+
+type AsyncUploader struct {
+	unit     *engine.Unit
+	uploader Uploader
+	log      zerolog.Logger
+	metrics  module.ExecutionMetrics
+}
+
+func (a *AsyncUploader) Ready() <-chan struct{} {
+	return a.unit.Ready()
+}
+
+func (a *AsyncUploader) Done() <-chan struct{} {
+	return a.unit.Done()
+}
+
+func (a *AsyncUploader) Upload(computationResult *execution.ComputationResult) error {
+	a.unit.Launch(func() {
+		a.metrics.ExecutionBlockDataUploadStarted()
+		defer a.metrics.ExecutionBlockDataUploadFinished()
+
+		err := a.uploader.Upload(computationResult)
+		if err != nil {
+			a.log.Error().Err(err).Msg("error while uploading block data")
+		}
+	})
+	return nil
+}
+
+func NewGCPBucketUploader(ctx context.Context, bucketName string, log zerolog.Logger) (*GCPBucketUploader, error) {
+
+	// no need to close the client according to documentation
+	ctx = context.Background()
+	client, err := storage.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create GCP Bucket client: %w", err)
+	}
+	bucket := client.Bucket(bucketName)
+
+	// try accessing buckets to validate settings
+	_, err = bucket.Attrs(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error while listing bucket attributes: %w", err)
+	}
+
+	return &GCPBucketUploader{
+		bucket: bucket,
+		log:    log.With().Str("subcomponent", "gcp_bucket_uploader").Logger(),
+		ctx:    ctx,
+	}, nil
+}
+
+type GCPBucketUploader struct {
+	log    zerolog.Logger
+	bucket *storage.BucketHandle
+	ctx    context.Context
+}
+
+func (u *GCPBucketUploader) Upload(computationResult *execution.ComputationResult) error {
+
+	objectName := GCPBlockDataObjectName(computationResult)
+	object := u.bucket.Object(objectName)
+
+	writer := object.NewWriter(u.ctx)
+	defer func() {
+		err := writer.Close()
+		if err != nil {
+			u.log.Warn().Err(err).Str("object_name", objectName).Msg("error while closing GCP object")
+		}
+	}()
+
+	return WriteComputationResultsTo(computationResult, writer)
+}
+
+func GCPBlockDataObjectName(computationResult *execution.ComputationResult) string {
+	return fmt.Sprintf("%s.cbor", computationResult.ExecutableBlock.ID().String())
```

Comment:
IIUC, this will silently overwrite the object.
1. do we want to document this in a comment?
2. do we want to consider object versioning in the case of multiple writers?
3. do we want to have GCPBucketUploader accept a constructor string prefix (that would allow e.g. a hot-swappable secondary) ?

Reply:
While all those comments are valid question, I think we should address them in next planned version.

Reply:
I think object versioning would be setup on the GCP bucket itself, so it wouldnt involve any changes here. GCS would automatically update the version when a new file of the same name is uploaded.